everestrs/everestrs/src/lib.rs

```diff
@@ -128,85 +133,106 @@ pub trait GenericModule {
     fn on_ready(&self) {}
 }
 
-#[derive(Clone)]
-pub struct RawPublisher {
-    cpp_module: cxx::SharedPtr<ffi::Module>,
+pub trait Publisher: Sync {
+    fn publish_variable<T: serde::Serialize>(&self, impl_id: &str, var_name: &str, message: &T);
+
+    fn call_command<T: serde::Serialize, R: serde::de::DeserializeOwned>(
+        &self,
+        impl_id: &str,
+        name: &str,
+        args: &T,
+    ) -> R;
 }
 
-impl RawPublisher {
-    pub fn publish_variable(&self, impl_id: &str, var_name: &str, data: &impl Serialize) {
+impl Publisher for Runtime {
+    fn publish_variable<T: serde::Serialize>(&self, impl_id: &str, var_name: &str, message: &T) {
         let blob = ffi::JsonBlob::from_vec(
-            serde_json::to_vec(data).expect("Serialization of data cannot fail."),
+            serde_json::to_vec(&message).expect("Serialization of data cannot fail."),
         );
         (self.cpp_module)
             .as_ref()
             .unwrap()
             .publish_variable(impl_id, var_name, blob);
     }
-   
-    pub fn call_command(&self, impl_id: &str, name: &str, args: &impl Serialize) -> serde_json::Value {
+
+    fn call_command<T: serde::Serialize, R: serde::de::DeserializeOwned>(
+        &self,
+        impl_id: &str,
+        name: &str,
+        args: &T,
+    ) -> R {
         let blob = ffi::JsonBlob::from_vec(
             serde_json::to_vec(args).expect("Serialization of data cannot fail."),
         );
         let return_value = (self.cpp_module)
             .as_ref()
             .unwrap()
             .call_command(impl_id, name, blob);
-        return_value.deserialize()
+        serde_json::from_slice(&return_value.data).unwrap()
     }
 }
 
 pub struct Runtime {
     // There are two subtleties here:
-    // 1. We are handing out pointers to `module_impl` to `cpp_module` for callbacks. The pointers
-    //    must must stay valid for as long as `cpp_module` is alive. Hence `module_impl` must never
+    // 1. We are handing out pointers to `sub_impl` to `cpp_module` for callbacks. The pointers
+    //    must must stay valid for as long as `cpp_module` is alive. Hence `sub_impl` must never
     //    move in memory. Rust can model this through the Pin concept which upholds this guarantee.
     //    We use a Box to put the object on the heap.
-    // 2. For the same reason, `module_impl` should outlive `cpp_module`, hence should be dropped
+    // 2. For the same reason, `sub_impl` should outlive `cpp_module`, hence should be dropped
     //    after it. Rust drops fields in declaration order, hence `cpp_module` should come before
-    //    `module_impl` in this struct.
-    cpp_module: cxx::SharedPtr<ffi::Module>,
-    module_impl: Pin<Box<dyn GenericModule>>,
+    //    `sub_impl` in this struct.
+    cpp_module: cxx::UniquePtr<ffi::Module>,
```

Comment:
this can't be a shared ptr for what we wrote above - the lifetime of the `cpp_module` must never exceed the life time of the `sub_impl`. Making it shared breaks this assumption

Reply:
Yes, that is right. Great that you found a fix.